Objects/tupleobject.c

```diff
@@ -333,39 +333,49 @@ tuplerepr(PyTupleObject *v)
     return NULL;
 }
 
-/* The addend 82520, was selected from the range(0, 1000000) for
-   generating the greatest number of prime multipliers for tuples
-   up to length eight:
 
-     1082527, 1165049, 1082531, 1165057, 1247581, 1330103, 1082533,
-     1330111, 1412633, 1165069, 1247599, 1495177, 1577699
+/* Hash for tuples. This is a slightly simplified version of the xxHash
+   non-cryptographic hash: we do not use any parallellism, there is only
+   1 accumulator. For the specification, see
+   https://github.com/Cyan4973/xxHash/wiki/xxHash-specification-(draft)
 
    Tests have shown that it's not worth to cache the hash value, see
-   issue #9685.
+   https://bugs.python.org/issue9685
 */
-
 static Py_hash_t
 tuplehash(PyTupleObject *v)
 {
-    Py_uhash_t x;  /* Unsigned for defined overflow behavior. */
-    Py_hash_t y;
-    Py_ssize_t len = Py_SIZE(v);
-    PyObject **p;
-    Py_uhash_t mult = _PyHASH_MULTIPLIER;
-    x = 0x345678UL;
-    p = v->ob_item;
-    while (--len >= 0) {
-        y = PyObject_Hash(*p++);
-        if (y == -1)
+    Py_uhash_t ERR = -1;
+    Py_ssize_t i, len = Py_SIZE(v);
+    PyObject **item = v->ob_item;
+
+    Py_uhash_t acc = _PyHASH_XXPRIME_5;
+    for (i = 0; i < len; i++) {
+        Py_uhash_t lane = PyObject_Hash(item[i]);
+        if (lane == ERR) {
             return -1;
-        x = (x ^ y) * mult;
-        /* the cast might truncate len; that doesn't change hash stability */
-        mult += (Py_hash_t)(82520UL + len + len);
+        }
+        acc += lane * _PyHASH_XXPRIME_2;
+        /* Rotate acc left by 13 or 31 bits */
+        acc = (acc << (sizeof(acc) <= 4 ? 13 : 31)) +
+              (acc >> (sizeof(acc) <= 4 ? 19 : 33));
```

Comment:
Here I think it would be better to #define a "rotate" macro at the place that already does different things depending on the size of `Py_uhash_t`, with constants hard-coded appropriate for the size in use.  The simpler the code, the more likely a compiler will emit a single rotate _instruction_ (e.g., I checked that Visual Studio compiled my spelling to a single rotate in the code I posted on BPO).

Reply:
Turns out Visual Studio does _not_ generate a rotate instruction for this, but for a different reason:  using "+" to combine the pieces.  Compilers do dumb pattern matching for stuff like this, and "everyone" writes a rotate by hand using "|" instead.  Use "|" instead of "+", and it does generate a single rotate instruction.